File: 8kyu/(8 kyu) Rock Paper Scissors/(8 kyu) Rock Paper Scissors.dart
// 1 Plain solution
String rps(String p1, String p2) {
  if (p1 == 'scissors' && p2 == 'paper') {
    return 'Player 1 won!';
  } else if (p1 == 'scissors' && p2 == 'rock') {
    return 'Player 2 won!';
  } else if (p1 == 'rock' && p2 == 'paper') {
    return 'Player 2 won!';
  } else if (p1 == 'rock' && p2 == 'scissors') {
    return 'Player 1 won!';
  } else if (p1 == 'paper' && p2 == 'rock') {
    return 'Player 1 won!';
  } else if (p1 == 'paper' && p2 == 'scissors') {
    return 'Player 2 won!';
  } else {
    return 'Draw!';
  }
}

// 2 Optimized solution
String rps(String p1, String p2) {
  if (p1 == p2) return 'Draw!';
  if (p1 == 'rock' && p2 == 'scissors') return 'Player 1 won!';
  if (p1 == 'scissors' && p2 == 'paper') return 'Player 1 won!';
  if (p1 == 'paper' && p2 == 'rock') return 'Player 1 won!';
  return 'Player 2 won!';
}

// 3 Clever solution
String rps(String p1, String p2) {
  return p1 == p2
      ? 'Draw!'
      : 'Player ${p1 == 'rock' && p2 == 'scissors' || p1 == 'scissors' && p2 == 'paper' || p1 == 'paper' && p2 == 'rock' ? 1 : 2} won!';
}

// 4 Coding golf
String rps(String p1, String p2) => p1 == p2
    ? 'Draw!'
    : 'Player ${'prsp'.contains(p1[0] + p2[0]) ? 1 : 2} won!';

File: 8kyu/(8 kyu) Rock Paper Scissors/(8 kyu) Rock Paper Scissors.py
# 1 Plain solution
def rps(p1, p2):
    if p1 == "scissors" and p2 == "paper":
        return "Player 1 won!"
    elif p1 == "scissors" and p2 == "rock":
        return "Player 2 won!"
    elif p1 == "rock" and p2 == "paper":
        return "Player 2 won!"
    elif p1 == "rock" and p2 == "scissors":
        return "Player 1 won!"
    elif p1 == "paper" and p2 == "rock":
        return "Player 1 won!"
    elif p1 == "paper" and p2 == "scissors":
        return "Player 2 won!"
    else:
        return "Draw!"

# 2 Optimized solution
def rps(p1, p2):
    if p1 == p2: return "Draw!"
    if p1 == "rock": return "Player 1 won!" if p2 == "scissors" else "Player 2 won!"
    if p1 == "paper": return "Player 1 won!" if p2 == "rock" else "Player 2 won!"
    if p1 == "scissors": return "Player 1 won!" if p2 == "paper" else "Player 2 won!"

# 3 Clever solution
def rps(p1, p2):
    if p1 == p2: return "Draw!"
    outcomes = {
        "rock": "scissors",
        "paper": "rock",
        "scissors": "paper"
    }
    return "Player 1 won!" if outcomes[p1] == p2 else "Player 2 won!"
